### PR TITLE
add column-based data structure and related unit test

### PR DESCRIPTION
**Pull request by Myrrolinz**
- bitmask
- cypher_string_t: new basic data structure for string type. For string type <= 4 bytes, store it in prefix[], otherwise store the rest of it in a buffer.
- column_vector: column-based structure. `ColumnVector` for int, `StringColumn` for string.
- test_column_vector: test file.

### CODE REVIEW

File: src/core/column_vector.h [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+    }
+
+    void CopyFromVectorData(uint8_t* dst_data, const ColumnVector* src_vector,
+                            const uint8_t* src_vector_data) {
+        std::memcpy(dst_data, src_vector_data, src_vector->GetElementSize());
+    }
+
+    void CopyFromVectorData(uint64_t dst_pos, const ColumnVector* src_vector, uint64_t src_pos) {
+        std::memcpy(data_.get() + dst_pos * element_size_, src_vector->data() + src_pos
+                    * element_size_, element_size_);
```
Comment (yannan-wyn) [inline src/core/column_vector.h:130]: 确认下这个复杂的copy函数的用途
我感觉暂时没用可以先删掉

File: src/core/column_vector.h [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+class ColumnVector {
+    friend class StringColumn;
+
+ public:
+    explicit ColumnVector(size_t element_size, size_t capacity = DEFAULT_VECTOR_CAPACITY)
+        : element_size_(element_size),
+          capacity_(capacity),
+          data_(new uint8_t[element_size * capacity]()),
+          bitmask_(capacity),
+          overflow_buffer_capacity_(1024),
```
Comment (yannan-wyn) [inline src/core/column_vector.h:37]: 不是每一个Vector都需要 overflow_buffer_

File: src/core/column_vector.h [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+
+    ColumnVector(const ColumnVector& other)
+        : element_size_(other.element_size_),
+          capacity_(other.capacity_),
+          data_(new uint8_t[other.element_size_ * other.capacity_]),
+          bitmask_(other.bitmask_),
+          overflow_buffer_capacity_(other.overflow_buffer_capacity_),
+          overflow_buffer_(new uint8_t[other.overflow_buffer_capacity_]),
+          overflow_offset_(other.overflow_offset_) {
+        std::copy(other.data_.get(), other.data_.get() + other.element_size_
```
Comment (yannan-wyn) [inline src/core/column_vector.h:49]: 这里为什么不用 std::memcpy 用的是 std::copy?

File: src/core/column_vector.h [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+        : element_size_(element_size),
+          capacity_(capacity),
+          data_(new uint8_t[element_size * capacity]()),
+          bitmask_(capacity),
+          overflow_buffer_capacity_(1024),
+          overflow_buffer_(new uint8_t[overflow_buffer_capacity_]),
+          overflow_offset_(0) {}
+
+    ColumnVector(const ColumnVector& other)
+        : element_size_(other.element_size_),
```
Comment (yannan-wyn) [inline src/core/column_vector.h:42]: 为何需要一个deep copy？

File: src/core/bit_mask.h [hunk 5]
```diff
@@ -0,0 +1,223 @@
+/**
```
[the remark this answers is not in the document]
Comment (Myrrolinz) [reply src/core/bit_mask.h:1]: ok

File: src/core/column_vector.h [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+        if (pos >= capacity_) {
+            throw std::out_of_range("Index out of range in GetValue");
+        }
+        reinterpret_cast<T*>(data_.get())[pos] = val;
+    }
+
+    void CopyFromRowData(uint32_t pos, const uint8_t* row_data) {
+        std::memcpy(data_.get() + pos * element_size_, row_data, element_size_);
+    }
+
```
[the remark this answers is not in the document]
Comment (Myrrolinz) [reply src/core/column_vector.h:118]: 用不到，已经删了

File: src/core/column_vector.h (hunk 3 above)
Comment (Myrrolinz) [reply src/core/column_vector.h:49]: 已经改成了memcpy

**Pull request closed without merging**